### ISSUE TITLE
Unable to assign /32 IPv4 addresses

### ISSUE THREAD

**Issue by bradmwalker** [mininet/mininet#675]
A Node specified with a /32 IPv4 address will have the netmask set to 0. This could be fixed by explicitly passing the netmask to ifconfig or using "ip addr", which correctly sets a /32 netmask.

**Comment by dmolavi** [comment]
I am also unable to assign /32 addresses:

`	# Add hosts and switches
	leftHost1  = self.addHost('h1',ip='10.0.0.1/32',mac='00:04:00:00:00:01')
	leftHost2  = self.addHost('h2',ip='10.0.0.2/32',mac='00:04:00:00:00:02')
	rightHost3 = self.addHost('h3',ip='10.0.0.3/32',mac='00:04:00:00:00:03')
	rightHost4 = self.addHost('h4',ip='10.0.0.4/32',mac='00:04:00:00:00:04')`

When I run this and do an `h1 ifconfig` from the mininet CLI, I get:

`mininet> h1 ifconfig
h1-eth1   Link encap:Ethernet  HWaddr 00:04:00:00:00:01  
          inet addr:10.0.0.1  Bcast:255.255.255.255  Mask:0.0.0.0
          inet6 addr: fe80::204:ff:fe00:1/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:22516 errors:0 dropped:5 overruns:0 frame:0
          TX packets:10 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1576195 (1.5 MB)  TX bytes:828 (828.0 B)
`
FWIW, /24 (and presumably others) work as expected.

**Comment by dmolavi** [comment]
It looks like the proper prefix length is being sent, however, ifconfig is barfing on the /32.  I can set everything up to /31 manually at the mininet CLI by doing something like:
`h1 ifconfig h1-eth1 10.0.0.1/31`

The resultant ifconfig shows the 255.255.255.254 masking.  However, if I do an:
`h1 ifconfig h1-eth1 10.0.0.1/32`

The ifconfig for h1 shows an all 0's mask.  

If I'm at my shell for my Ubuntu (15.10) machine and do an:
`ifconfig eth0 10.1.1.1/32`

It shows an all 0's mask as well.  This is not a mininet problem, IMHO, and can be closed.

**Comment by dmolavi** [comment]
Upon further inspection, I propose that mininet use the "ip" command to set addresses vs the "ifconfig" command.  I can enter:
`sudo ip addr add dev eth0 10.0.0.1/32`
at my Ubuntu command line, and the resultant mask appears as all 255, as expected.

**Comment by lantz** [comment]
I agree - `ifconfig` should do the right thing! Changing to `ip route` brings with it a number of issues - for example I'm not sure that `ip addr add` actually replaces an existing IP address... Nonetheless, I'd like to see these sorts of things addressed, along with IPv6 support as well. If you'd like to start taking a crack at it, I'd be interested in your thoughts or suggestions.

There is an obvious workaround of course - you can simply call `ip a add` manually.

**Comment by dmolavi** [comment]
`ip address change...` does not modify the address on the device you specify, contrary to what the man page states.  So using the `ip` command would only be good for initial interface configuration, but not modification.  

I'll do some more digging and see if there's some other way to get a /32 specified with ifconfig.

**Comment by dmolavi** [comment]
`ifconfig` will properly set a /32 if you use the following syntax:
`ifconfig eth0 10.10.10.10 netmask 255.255.255.255`

I think the least invasive fix to this would be to add a quick check where the prefix length is set, if it's a /32 use the above syntax, else use the existing syntax.

**Comment by bradmwalker** [comment]
@dmolavi Why not always generate the netmask w/ifconfig?

**Comment by dmolavi** [comment]
Because for anything other than 8, 16, and 24, you'd have to do some sort of lookup in a dict. I'm trying to keep the overall code impact to a minimum and just make the one change for a /32.

**Comment by dmolavi** [comment]
Thanks to the power of Google, I found this:
http://code.activestate.com/recipes/576483-convert-subnetmask-from-cidr-notation-to-dotdecima/ 

See David Moss' reply with this code snip:
`from socket import inet_ntoa
from struct import pack

def calcDottedNetmask(mask):
    bits = 0
    for i in xrange(32-mask,32):
        bits |= (1 << i)
    return inet_ntoa(pack('>I', bits))`

This could be used to return a proper netmask and use the `ifconfig` syntax from above.  I'll defer to @lantz ...

**Comment by bradmwalker** [comment]
>>> import ipaddress
>>> address, netmask = ipaddress.ip_interface('192.100.0.2/19').with_netmask.split('/')
>>> address, netmask
('192.100.0.2', '255.255.224.0')

**Comment by dmolavi** [comment]
I had no idea that was available. And I'm kind of surprised it didn't appear on google.  But yeah, +1 on that one.

**Comment by dmolavi** [comment]
Hmm that appears to be a Python 3.3+ function only...it fails on my 2.7 install. 

://docs.python.org/3/library/ipaddress.html
